source-syntax/scribblings/source-syntax.scrbl

```diff
@@ -0,0 +1,23 @@
+#lang scribble/manual
+@require[scribble/example (for-label racket/base racket/contract)]
+
+@title{Source Syntax}
+
+@defmodule[syntax/source-syntax]{}
+
+@defproc[(recover-source-syntax [orig syntax?]
+                                [expanded syntax?]
+                                [#:traverse-now? now? boolean? #f])
+         (-> syntax? (or/c syntax? #f))]{
+  Return a procedure that accepts a syntax object from @racket[expanded]
+  and returns the outermost syntax object in @racket[orig] that has the same
+  location as the given syntax object.
```

Comment:
I forget the exact details (@samth would know), but I think this is only the best-case scenario.
I think it's sometimes necessary to fall back to an enclosing syntax object.

Reply:
What do you mean by "enclosing"?

(If `A` and `B` are syntax objects in `orig` with the same location `loc` and `A` contains `B`, then a lookup for `loc` should return `A` (aka the enclosing object))

Reply:
I think the tricky case arises when no original syntax object has location `loc`. In which case (IIRC) the library tries to find the tightest original syntax object whose location surrounds `loc`. But I forget the details.

Reply:
Ok, looks like:
- if you search with a syntax object `e` with location `loc`
- and no original syntax object has location `loc`
- then you get the result for the nearest ancestor of `e` that has a location in the original syntax

Reply:
That sounds right.
It's a bit ad-hoc, but I think it's helped get better error messages for TR.
In any case, worth documenting, I think.
Thanks for digging it up!